### PR TITLE
BRMS 45 (add utility for providing TLS certificate expiry dates status)

### PR DESCRIPTION
**Pull request by chris-ianson**
@adamconder 

I've added a CertificateStatus class to help us track the expiry date of a certificate. 

It contains the following methods...

_getExpiryDate
difference
statusMessage
isValidDate_

The way it works out the difference between two dates is open to change, I'm not sure how you've done it in the past and if you agree with returning touples (the method returns day, month year ints)

statusMessage() uses the difference response to build a message that we can log.

isValidDate checks to see if the difference in days is positive and can be called to determine if the certificate has a valid date

Happy to update the code if you've done something different or you think we should tweak it. We still have to determine where in the app we'll log to Kibana but hopefully this class can help us create the log messages.

### CODE REVIEW

File: app/uk/gov/hmrc/brm/utils/CertificateStatus.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+  lazy val privateKeystoreKey = getConfString(privateKeystoreKey_Key,
+    throw new RuntimeException("[Configuration][NotFound] privateKeystoreKey"))
+
+  val certificateExpiryDate_Key = "birth-registration-matching.certificateExpiryDate"
+
+  lazy val certificateExpiryDate = getConfString(certificateExpiryDate_Key,
+    throw new RuntimeException("[Configuration][NotFound] certificateExpiryDate"))
+
+  def getExpiryDate(expiryDate: Option[String] = None): LocalDate = expiryDate match {
```
Comment (adamconder) [inline app/uk/gov/hmrc/brm/utils/CertificateStatus.scala]: Change the signature of this method to be:
`def getExpiryDate(expiryDate: Option[String] = Some(certificateExpiryDate) : LocalDate`

File: app/uk/gov/hmrc/brm/utils/CertificateStatus.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    case None => new LocalDate(certificateExpiryDate)
```
Comment (adamconder) [inline app/uk/gov/hmrc/brm/utils/CertificateStatus.scala]: Pass in the default `expiryDate` as the argument here.

File: app/uk/gov/hmrc/brm/utils/CertificateStatus.scala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    case (0, 0, 0) => "expires today"
+    case (d, 0, 0) if d <= 31 => "expires this month"
+    case (d, m, 0) if (m <= 12 && m > 0) => s"expires in $m months, $d days"
+    case (d, m, y) if (y > 0) => s"expires in $y years, $m months, $d days"
+    case _ => "couldn't determine status message from given dates"
+  }
+
+  def isValidDate(): Boolean = {
+    val dateDifference = difference(getExpiryDate(), new LocalDate())
+    if(dateDifference._1 > 0) {
```
Comment (adamconder) [inline app/uk/gov/hmrc/brm/utils/CertificateStatus.scala]: Change to `dateDifference._! >0`

Comment (adamconder) [inline app/uk/gov/hmrc/brm/utils/CertificateStatus.scala]: Change this to be:

```
val (day, month, year) = difference(getExpiryDate(), new LocalDate())
logCertificateExpiry(day, month, year)
dateDifference._1 > 0
```

File: app/uk/gov/hmrc/brm/utils/CertificateStatus.scala [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+  def difference(expiryDate: LocalDate, comparisonDate: LocalDate): (Int, Int, Int) = {
+    val days = Days.daysBetween(comparisonDate, expiryDate).getDays
+    val months = Months.monthsBetween(comparisonDate, expiryDate).getMonths
+    val years = Years.yearsBetween(comparisonDate, expiryDate).getYears
+    (days, months, years)
+  }
+
+  def statusMessage(dayDifference: Int, monthDifference: Int, yearDifference: Int): String = (dayDifference, monthDifference, yearDifference) match {
+    case (0, 0, 0) => "expires today"
```
Comment (adamconder) [inline app/uk/gov/hmrc/brm/utils/CertificateStatus.scala]: `case (0, 0, 0) => Logger.error("[GROProxy][Certificate][EXPIRES_TODAY]")`

File: app/uk/gov/hmrc/brm/utils/CertificateStatus.scala [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    case (d, 0, 0) if d <= 31 => "expires this month"
```
Comment (adamconder) [inline app/uk/gov/hmrc/brm/utils/CertificateStatus.scala]: `case (d, 0, 0) if d <= 31 => Logger.error("[GROProxy][Certificate][EXPIRES_THIS_MONTH]")`

File: app/uk/gov/hmrc/brm/utils/CertificateStatus.scala [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    case (d, m, 0) if (m <= 12 && m > 0) => s"expires in $m months, $d days"
```
Comment (adamconder) [inline app/uk/gov/hmrc/brm/utils/CertificateStatus.scala]:
```
case (d, m, 0) if (m <= 12 && m > 0) => 
if (m <= 6) {
Logger.warn("[GROProxy][Certificate][EXPIRES_IN][Days $d][Months: $m]")
} else {
Logger.info("[GROProxy][Certificate][EXPIRES_IN][Days $d][Months: $m]")
}

```

File: app/uk/gov/hmrc/brm/connectors/BirthConnector.scala [hunk 7]
```diff
@@ -125,7 +126,10 @@ trait BirthConnector extends ServicesConfig {
     )
   }
 
-  private def requestAuth(body: BirthResponse => BirthResponse)(implicit hc: HeaderCarrier) = {
+  def requestAuth(body: BirthResponse => BirthResponse)(implicit hc: HeaderCarrier) = {
```
Comment (manishsw) [inline app/uk/gov/hmrc/brm/connectors/BirthConnector.scala]: kindly change it to private.

File: app/uk/gov/hmrc/brm/connectors/BirthConnector.scala [hunk 8]
```diff
@@ -150,7 +153,7 @@ trait BirthConnector extends ServicesConfig {
     body(handleResponse(response, extractAccessToken, "requestAuth"))
   }
 
-  private def requestReference(reference: String)(implicit hc: HeaderCarrier): BirthResponse = {
+  def requestReference(reference: String)(implicit hc: HeaderCarrier): BirthResponse = {
```
Comment (manishsw) [inline app/uk/gov/hmrc/brm/connectors/BirthConnector.scala]: make it private.

File: app/uk/gov/hmrc/brm/utils/CertificateStatus.scala [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+  private def difference(expiryDate: LocalDate, comparisonDate: LocalDate): (Int, Int, Int) = {
+    val days = Days.daysBetween(comparisonDate, expiryDate).getDays
+    val months = Months.monthsBetween(comparisonDate, expiryDate).getMonths
+    val years = Years.yearsBetween(comparisonDate, expiryDate).getYears
+    (days, months, years)
+  }
+
+  private def logCertificate(dayDifference: Int, monthDifference: Int, yearDifference: Int): Unit = (dayDifference, monthDifference, yearDifference) match {
+    case (0, 0, 0) => Logger.error("[GROProxy][Certificate][EXPIRES_TODAY]")
```
Comment (manishsw) [inline app/uk/gov/hmrc/brm/utils/CertificateStatus.scala]: use BRM logger.

File: test/utils/CertificateStatusSpec.scala [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+      dateStatus shouldBe true
+    }
+
+    "return true when current date 10 months behind certificate expiry date" in {
+      val dateStatus = mockCertificateStatus20160219.logCertificateStatus(new LocalDate("2016-02-19").minusMonths(10: Int))
+      dateStatus shouldBe true
+    }
+
+    "return false when current date 10 months in front of certificate expiry date" in {
+      val dateStatus = mockCertificateStatus20160219.logCertificateStatus(new LocalDate("432016-02-19").plusMonths(10: Int))
```
Comment (manishsw) [inline test/utils/CertificateStatusSpec.scala]: may required valid date.

**Pull request merged**